### ISSUE TITLE
Error when hashing files on .NET6

### ISSUE THREAD

**Issue by EvgeniyZ-ru** [alanmcgovern/monotorrent#483]
When checking the files (hashing) I get the following error:

```
ReadFailure
Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Unknown error (0xc1000008)
   at Internal.Cryptography.HashProviderCng.AppendHashData(ReadOnlySpan`1 source)
   at Internal.Cryptography.HashProvider.AppendHashData(Byte[] data, Int32 offset, Int32 count)
   at MonoTorrent.Client.DiskManager.GetHashAsync(ITorrentData manager, Int32 pieceIndex)
   at MonoTorrent.Client.DiskManager.GetHashAsync(ITorrentData manager, Int32 pieceIndex)
   at System.Runtime.CompilerServices.ReusableTaskAwaiter`1.GetResult()
   at MonoTorrent.Client.Modes.HashingMode.WaitForHashingToComplete()
   at MonoTorrent.Client.TorrentManager.HashCheckAsync(Boolean autoStart, Boolean setStoppedModeWhenDone)
```

This happens when I delete one of files, hoping it will re-download. After a little searching for the cause of the problem, I came across [this](https://github.com/dotnet/runtime/issues/61417) where a person had a similar problem (with another program) on .NET6. I checked it and yes, on .NET5 the error does not occur.

**Comment by alanmcgovern** [comment]
Oh interesting! I'll have to see if there's a threading issue there. Two threads should definitely not be using that concurrently.

**Comment by alanmcgovern** [comment]
@EvgeniyZ-ru Which version of the nuget are you using, or which revision of the repository are you compiling?

**Comment by EvgeniyZ-ru** [comment]
@alanmcgovern latest, 2.0.1 (NuGet)

**Comment by alanmcgovern** [comment]
@EvgeniyZ-ru Hrmm, I can't find anything which indicates concurrent access to the SHA1 hasher.

My testcase was to download part of a ubuntu iso torrent, and then put the sample client in an infinite loop calling 'HashCheckASync'.  This puts it in 'HashingMode', just like your stacktrace implies.

Perhaps i need a newer version of .NET6 to trigger this, in which case it's most likely a bug in the framework itself.

However, to help rule out a bug could you compile this vesrion of monotorrent locally and see if it indicates a double-use issue?  https://github.com/alanmcgovern/monotorrent/commit/8f09db406c93b2edc994e27b4d259e737bca74b3

It puts a check on the 'ReusableExclusiveSemaphore' to ensure double frees never occur. I'd be shocked if this were the issue, but it's the most plausible thing other than a .NET6 framework bug.

**Comment by alanmcgovern** [comment]
[MonoTorrent.2.0.0-alpha.unstable.rev0004.zip](https://github.com/alanmcgovern/monotorrent/files/7649898/MonoTorrent.2.0.0-alpha.unstable.rev0004.zip)

If it helps, you can use this unstable/test nuget. It's built from the branch I linked in my previous comment :)

**Comment by EvgeniyZ-ru** [comment]
@alanmcgovern Unfortunately, this did not help.
I replaced NuGet package with the one in the archive, rebuilt the application, launched it but the error remained. After that, I thought that the problem with my application, downloaded the sources (the branch, that you have indicated), built `SimpleClient` (replaced there only package with the one that you have provided in the archive, because I could not build), put in `Torrents` directory the file I need, and in `Downloads` previously downloaded files (without 1, which I myself had removed for the test) and run.
The result, the following:

```
while (true) {
    await manager.HashCheckAsync (false);
    Console.WriteLine ("Done");
}
```
After a long time, gave out `Done` (70gb torrent). I commented out this part of the code to make it work properly, hashing starts, but at a certain point error again (status Error).

`manager.Error.Reason`: ReadFailure
`manager.Error.Message`: Unknown error (0xc1000008)
`manager.Error.StackTrace`:
```
   в Internal.Cryptography.HashProviderCng.AppendHashData(ReadOnlySpan`1 source)
   в Internal.Cryptography.HashProvider.AppendHashData(Byte[] data, Int32 offset, Int32 count)
   в MonoTorrent.Client.DiskManager.<GetHashAsync>d__57.MoveNext()
   в System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   в MonoTorrent.Client.DiskManager.<GetHashAsync>d__57.MoveNext()
   в System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   в System.Runtime.CompilerServices.ReusableTaskAwaiter`1.GetResult()
   в MonoTorrent.Client.Modes.HashingMode.<WaitForHashingToComplete>d__15.MoveNext()
   в System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   в System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   в System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   в MonoTorrent.Client.TorrentManager.<HashCheckAsync>d__183.MoveNext()
```

`MonoTorrent Version`: MonoTorrent.2.0.0-alpha.unstable.rev0004
`Source Code`: [this](https://github.com/alanmcgovern/monotorrent/tree/almcgo/test-issue-483/src/SampleClient)
`dotnet --list-sdks`:  6.0.100 [C:\Program Files\dotnet\sdk]
`OS`: Windows 11 (Windows 10 too)

**Comment by alanmcgovern** [comment]
Hrm, that's frustrating. I had a 2-3GB torrent hashing in a loop and it survived about 30-50 iterations with no problems. 

Can you describe exactly what you're doing, or share a small program which I can compile and use to reproduce the problem?

For example, are you creating a ClientEngine, then adding one Torrent, then calling StartAsync, and then it crashes while doing the initial hash check? Or are you adding it and calling HashCheckAsync?

Or are you adding and hashing more than 1 torrent at the same time, or adding many in a row? 

It's confusing as (despite appearances) HashingMode is essentially single threaded. It's super surprising to me that there could be a race there, even if there are more than 1 torrents loaded in the engine. 

Are you using ClientEngine.DownloadMetadataAsync, or ClientEngine.StartAllAsync while adding/starting/hashing torrents manually?

Is there anything you're doing which could cause two hashchecks to happen at the same time?

This is quite perplexing :)

**Comment by EvgeniyZ-ru** [comment]
My goal is to create an application that will download one particular handout based on a hash string and then check on request the integrity of the files. 

While testing your library, I sketched this code:

```
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Humanizer;
using MonoTorrent;
using MonoTorrent.Client;

namespace ConsoleApp_net
{
    static class StopWatchUtils
    {
        public static TimeSpan GetEta(this Stopwatch sw, long counter, long counterGoal)
        {
            if (counter == 0) return TimeSpan.Zero;
            float elapsedMin = ((float)sw.ElapsedMilliseconds / 1000) / 60;
            float minLeft = (elapsedMin / counter) * (counterGoal - counter);
            TimeSpan ret = TimeSpan.FromMinutes(minLeft);
            return ret;
        }
    }

    class StandardDownloader
    {
        ClientEngine Engine { get; }

        public StandardDownloader(ClientEngine engine)
        {
            Engine = engine;
        }

        public async Task DownloadAsync(MagnetLink magnet, CancellationToken token)
        {
            var downloadsPath = Path.Combine(Environment.CurrentDirectory, "Downloads");

            var manager = await Engine.AddAsync(magnet, downloadsPath);
            await manager.StartAsync();

            var sb = new StringBuilder();
            var sw = Stopwatch.StartNew();

            while (Engine.IsRunning)
            {
                Console.SetCursorPosition(0, 0);
                sb.Clear();
                var peersCount = (await manager.GetPeersAsync()).Count;
                sb.AppendLine($"Hash: {manager.InfoHash.ToHex().ToLower()}");
                sb.AppendLine($"State: {manager.State} [DHT: {manager.CanUseDht}] [LPD: {manager.CanUseLocalPeerDiscovery}]");
                sb.AppendLine($"Name: {(manager.Torrent == null ? "MetaDataMode" : manager.Torrent.Name)}");
                sb.AppendLine($"Connections: [open:{Engine.ConnectionManager.OpenConnections} / {Engine.Settings.MaximumConnections}] [half-open:{Engine.ConnectionManager.HalfOpenConnections} / {Engine.Settings.MaximumHalfOpenConnections}]");
                sb.AppendLine($"Peers: {peersCount} / {manager.Peers.Available} [Seed: {manager.Peers.Seeds}] [Leechs: {manager.Peers.Leechs}]");
                sb.AppendLine($"Trackers: {manager.TrackerManager.Tiers.Count}");
                foreach (var tier in manager.TrackerManager.Tiers)
                    sb.AppendLine($"\t{tier.ActiveTracker} : Announce Succeeded: {tier.LastAnnounceSucceeded}. Scrape Succeeded: {tier.LastScrapeSucceeded}.");
                sb.AppendLine($"Progress: {manager.Progress:0.00}% [ETA: {sw.GetEta(manager.Monitor.DataBytesDownloaded, manager.Size).Humanize()}]");
                sb.AppendLine($"Speed: [d: {Engine.TotalDownloadSpeed.Bytes().Per(1.Seconds()).Humanize("#.##")}] [u: {Engine.TotalUploadSpeed.Bytes().Per(1.Seconds()).Humanize("#.##")}]");
                if (manager.PieceManager != null)
                    sb.AppendLine($"Pieces requested: {await manager.PieceManager.CurrentRequestCountAsync()} / {manager.PieceCount()}");


                if (manager.State == TorrentState.Error)
                {
                    sb.AppendLine(manager.Error.Reason.ToString());
                    sb.AppendLine(manager.Error.Exception.ToString());
                }

                Console.Write(sb);
                await Task.Delay(5000, token);
            }
        }
    }

    class Program
    {
        static async Task Main(string[] args)
        {
            CancellationTokenSource cancellation = new();

            Console.CancelKeyPress += (_, _) => cancellation.Cancel();
            AppDomain.CurrentDomain.ProcessExit += (_, _) => cancellation.Cancel();
            AppDomain.CurrentDomain.UnhandledException += (_, e) => { Console.WriteLine(e.ExceptionObject); cancellation.Cancel(); };
            Thread.GetDomain().UnhandledException += (_, e) => { Console.WriteLine(e.ExceptionObject); cancellation.Cancel(); };

            await Run(cancellation.Token);
            Console.ReadLine();
        }

        public static async Task Run(CancellationToken token)
        {
            Console.Write("Hash: ");
            var hashString = Console.ReadLine();

            var settings = new EngineSettingsBuilder()
            {
                MaximumOpenFiles = 50
            }.ToSettings();

            using var engine = new ClientEngine(settings);

            var hash = InfoHash.FromHex(hashString);
            var torrent = new MagnetLink(hash);

            try
            {
                var manager = new StandardDownloader(engine);
                await manager.DownloadAsync(torrent, token);
            }
            catch (OperationCanceledException)
            {

            }

            await engine.StopAllAsync();
            Console.WriteLine("Done...");
        }
    }
}

```

On startup it should ask for a hash string (you can take any hash string where there are several files), e.g. this `A616D9E907EF71F4472D74A9FE29E4B300CAA442` (music album 1.16 GB).

How I reproduce the problem:
- I start it up and specify the hash string.
- I am waiting for the download to complete.
- In the `Downloads\{distribution}` folder I remove any of the files.
- Also delete the folder `cache`.
- Re-run the program and enter the same hash.
- Starts the process of hashing the files.
- After some time, error.

Today I also checked the code you have in `SimpleClient`, just ran it and did the same thing (except I put the torrent). The result is the same.

Problem demonstration: [Imgur](https://imgur.com/AgnFBbF)

**Comment by alanmcgovern** [comment]
Yup, it's a framework bug!

This is the reproduction case:
```
            var buffer = new byte[1024];
            var hasher = SHA1.Create ();
            hasher.TransformBlock (buffer, 0, buffer.Length, buffer, 0);
            hasher.Initialize ();
            hasher.TransformBlock (buffer, 0, buffer.Length, buffer, 0);
            hasher.TransformFinalBlock (Array.Empty<byte> (), 0, 0);
            GC.KeepAlive (hasher.Hash);
            hasher.Initialize ();
            hasher.TransformBlock (buffer, 0, buffer.Length, buffer, 0)
```

If you initialize the hasher *after* transforming a block, and *before* transforming the final block, it breaks!

**Comment by alanmcgovern** [comment]
[MonoTorrent.2.0.2.zip](https://github.com/alanmcgovern/monotorrent/files/7653856/MonoTorrent.2.0.2.zip)

@EvgeniyZ-ru I've attached a 2.0.2 release. Could you double check it also solves the problem for you? The main change is this: https://github.com/alanmcgovern/monotorrent/commit/fbfd4b7a0c39ab063f39fa5390b3a5d99bf95a46 . There are also 2 patches improving how the DhtEngine handles the situation where it fails to bind to the correct port, but those shouldn't impact you!

If it's easier for you to compile form source, that change is in the monotorrent-2.0 branch too :)

If you're testing shows this works, I'll publish that binary to nuget.org

**Comment by EvgeniyZ-ru** [comment]
@alanmcgovern I tested this version by downloading 3 torrents followed by hashing them, also tested on Windows 10. No error appears, no matter how much I mock the program ;)

Thanks so much for the fix!

**Comment by alanmcgovern** [comment]
Great!

Thanks for the bug report too. I've released 2.0.2 with this fix. Hopefully it's the only bug/regression in .NET 6.0!
